// @aws-sdk/client-dynamodb/dist-es/commands/BatchGetItemCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { BatchGetItemInput, BatchGetItemOutput } from "../models/models_0";
import { deserializeAws_json1_0BatchGetItemCommand, serializeAws_json1_0BatchGetItemCommand, } from "../protocols/Aws_json1_0";
var BatchGetItemCommand = (function (_super) {
    __extends(BatchGetItemCommand, _super);
    function BatchGetItemCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    BatchGetItemCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "DynamoDBClient";
        var commandName = "BatchGetItemCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: BatchGetItemInput.filterSensitiveLog,
            outputFilterSensitiveLog: BatchGetItemOutput.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    BatchGetItemCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_0BatchGetItemCommand(input, context);
    };
    BatchGetItemCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_0BatchGetItemCommand(output, context);
    };
    return BatchGetItemCommand;
}($Command));
export { BatchGetItemCommand };
