// @aws-sdk/client-dynamodb/dist-es/commands/DescribeTableCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DescribeTableInput, DescribeTableOutput } from "../models/models_0";
import { deserializeAws_json1_0DescribeTableCommand, serializeAws_json1_0DescribeTableCommand, } from "../protocols/Aws_json1_0";
var DescribeTableCommand = (function (_super) {
    __extends(DescribeTableCommand, _super);
    function DescribeTableCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DescribeTableCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "DynamoDBClient";
        var commandName = "DescribeTableCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DescribeTableInput.filterSensitiveLog,
            outputFilterSensitiveLog: DescribeTableOutput.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DescribeTableCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_0DescribeTableCommand(input, context);
    };
    DescribeTableCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_0DescribeTableCommand(output, context);
    };
    return DescribeTableCommand;
}($Command));
export { DescribeTableCommand };
