// @aws-sdk/client-dynamodb/dist-es/commands/TransactWriteItemsCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { TransactWriteItemsInput, TransactWriteItemsOutput } from "../models/models_0";
import { deserializeAws_json1_0TransactWriteItemsCommand, serializeAws_json1_0TransactWriteItemsCommand, } from "../protocols/Aws_json1_0";
var TransactWriteItemsCommand = (function (_super) {
    __extends(TransactWriteItemsCommand, _super);
    function TransactWriteItemsCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    TransactWriteItemsCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "DynamoDBClient";
        var commandName = "TransactWriteItemsCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: TransactWriteItemsInput.filterSensitiveLog,
            outputFilterSensitiveLog: TransactWriteItemsOutput.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    TransactWriteItemsCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_0TransactWriteItemsCommand(input, context);
    };
    TransactWriteItemsCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_0TransactWriteItemsCommand(output, context);
    };
    return TransactWriteItemsCommand;
}($Command));
export { TransactWriteItemsCommand };
