// @aws-sdk/client-dynamodb/dist-es/commands/UpdateTableCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { UpdateTableInput, UpdateTableOutput } from "../models/models_0";
import { deserializeAws_json1_0UpdateTableCommand, serializeAws_json1_0UpdateTableCommand, } from "../protocols/Aws_json1_0";
var UpdateTableCommand = (function (_super) {
    __extends(UpdateTableCommand, _super);
    function UpdateTableCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    UpdateTableCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "DynamoDBClient";
        var commandName = "UpdateTableCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: UpdateTableInput.filterSensitiveLog,
            outputFilterSensitiveLog: UpdateTableOutput.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    UpdateTableCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_0UpdateTableCommand(input, context);
    };
    UpdateTableCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_0UpdateTableCommand(output, context);
    };
    return UpdateTableCommand;
}($Command));
export { UpdateTableCommand };
