// @aws-sdk/client-dynamodb/dist-es/commands/UpdateTimeToLiveCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { UpdateTimeToLiveInput, UpdateTimeToLiveOutput } from "../models/models_0";
import { deserializeAws_json1_0UpdateTimeToLiveCommand, serializeAws_json1_0UpdateTimeToLiveCommand, } from "../protocols/Aws_json1_0";
var UpdateTimeToLiveCommand = (function (_super) {
    __extends(UpdateTimeToLiveCommand, _super);
    function UpdateTimeToLiveCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    UpdateTimeToLiveCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "DynamoDBClient";
        var commandName = "UpdateTimeToLiveCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: UpdateTimeToLiveInput.filterSensitiveLog,
            outputFilterSensitiveLog: UpdateTimeToLiveOutput.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    UpdateTimeToLiveCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_0UpdateTimeToLiveCommand(input, context);
    };
    UpdateTimeToLiveCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_0UpdateTimeToLiveCommand(output, context);
    };
    return UpdateTimeToLiveCommand;
}($Command));
export { UpdateTimeToLiveCommand };
